aioesphomeapi/_frame_helper.py

```diff
@@ -219,21 +220,29 @@ def __init__(
     ) -> None:
         """Initialize the API frame helper."""
         super().__init__(on_pkt, on_error)
-        self._ready_event = asyncio.Event()
+        self._ready_future = asyncio.get_event_loop().create_future()
         self._noise_psk = noise_psk
         self._expected_name = expected_name
         self._state = NoiseConnectionState.HELLO
         self._setup_proto()
 
+    def _set_ready_future_exception(self, exc: Exception) -> None:
+        if not self._ready_future.done():
+            self._ready_future.set_exception(exc)
+
     def close(self) -> None:
         """Close the connection."""
         # Make sure we set the ready event if its not already set
         # so that we don't block forever on the ready event if we
         # are waiting for the handshake to complete.
-        self._ready_event.set()
+        self._set_ready_future_exception(APIConnectionError("Connection closed"))
         self._state = NoiseConnectionState.CLOSED
         super().close()
 
+    def _handle_error_and_close(self, exc: Exception) -> None:
+        self._set_ready_future_exception(exc)
```

Comment:
This will ensure the exception in the protocol callback raises upwards during the handshake